### PR TITLE
Add GPU support for ONNXRuntime

### PR DESCRIPTION
**Pull request by hqucms**
#### PR description:

This is a technical PR that allows the GPU support of ONNXRuntime (https://github.com/cms-sw/cmsdist/pull/6776) to be easily enabled in CMSSW.

Example usage:

```c++
  using namespace cms::Ort;
  std::string model_path = edm::FileInPath("PhysicsTools/ONNXRuntime/test/data/model.onnx").fullPath();
  auto session_options = ONNXRuntime::defaultSessionOptions(/*use_cuda=*/true);
  ONNXRuntime rt(model_path, &session_options);
```

#### PR validation:

Unit tests are updated to cover all three options.

### CODE REVIEW

**Comment by tvami** [comment]
@cmsbuild , please test

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc [hunk 1]
```diff
@@ -78,6 +77,35 @@ namespace cms::Ort {
...
+        options.cuda_mem_limit = std::numeric_limits<std::size_t>::max();
+        options.cudnn_conv_algo_search = OrtCudnnConvAlgoSearch::EXHAUSTIVE;
+        options.do_copy_in_default_stream = 1;
+        sess_opts.AppendExecutionProvider_CUDA(options);
+      } else {
+        // if GPU is not available
+        if (gpu_mode == force_gpu) {
+          throw cms::Exception("RuntimeError") << "No GPU detected, cannot run ONNXRuntime on GPU.";
+        } else {
+          std::cout << "[ONNXRuntime] No GPU detected, will run on CPU." << std::endl;
```
Comment (fwyzard) [inline PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:102]: please don't use `std::cout`, as it cannot be silenced from the job configuration

**Comment by hqucms** [comment]
> * how complicated would it be to extend to the Async mode in CMSSW?

I have never tried to use the async mode in CMSSW -- can you point me to any documentation or example?

> * now it grabs the whole GPU, how difficult would it be to extend to use streams?

It does not grabs the whole GPU, so in principle it can be used in multiple streams simultaneously I think (to be tested of course).

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc [hunk 2]
```diff
@@ -104,6 +132,10 @@ namespace cms::Ort {
       } else {
         input_dims = input_shapes[input_pos];
         // rely on the given input_shapes to set the batch size
+        if (input_dims[0] != batch_size) {
+          throw cms::Exception("RuntimeError") << "The first element of `input_shapes` (" << input_dims[0]
+                                               << ") does not match the given `batch_size` (" << batch_size << ")";
+        }
```
Comment (fwyzard) [inline PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:138]: this change is unrelated to the use of GPUs, isn't it ?

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 1 above)
Comment (hqucms) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:102]: Would `edm::LogInfo` work?

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 2 above)
Comment (hqucms) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:138]: No it's unrelated to GPU -- just a helpful error message for users.

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 1 above)
Comment (fwyzard) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:102]: Sure, or maybe `edm::LogWarning` if it's supposed to be a more visible message.

**Comment by fwyzard** [comment]
What NVIDIA GPU architectures (SM 6.0, 7.0, 7.5, _etc._) does ONNX support ?
Is it tied to how we use CUDA in CMSSW, or is it independent from it ?

**Comment by fwyzard** [comment]
> > * how complicated would it be to extend to the Async mode in CMSSW?
> 
> I have never tried to use the async mode in CMSSW -- can you point me to any documentation or example?

It looks like most (all ?) CMSSW modules that use ONNX are already `stream` modules and share the same `ONNXRuntime` instance,_e.g._:
```c++
class BoostedJetONNXJetTagsProducer : public edm::stream::EDProducer<edm::GlobalCache<ONNXRuntime>> {
  ...
```

**Comment by makortel** [comment]
Does ONNX itself provide an asynchronous API for inference?

**Comment by makortel** [comment]
In addition to what Andrea wrote in https://github.com/cms-sw/cmssw/pull/36963#issuecomment-1040314622, the documentation of `ExternalWork` extension is documented in
https://twiki.cern.ch/twiki/bin/view/CMSPublic/FWMultithreadedFrameworkStreamModuleInterface#edm_ExternalWork

**Comment by fwyzard** [comment]
The main concern I have is about the interface. Figuring out something more flixible now could avoid making changes later, but it could be that's easier to take the changes like this, and improve on the interface later on, once we have more experience with ONNX on GPUs.

Anyway
  - we shouldn't use "GPU" to mean "only an NVIDIA GPU"; maybe use "CUDA" instead ?
  - the choice to use of GPUs (yes, no, auto) should be made consistently with the rest of CMSSW, as in #36699; @makortel what do you think ?
  - it might actually be useful to expose the setting about the `options.cuda_mem_limit`, to make it easier to share the GPU with other modules in the same job, but also other jobs running on the same machine
  - does `device_id = 0` tell ONXX to only use the first available GPU ? this is probably OK to start with, but later on we need to figure out how to deal with multiple GPUs, and how pass this information to the ONNX runtime from the framework

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc [hunk 3]
```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
...
+    SessionOptions sess_opts;
+    sess_opts.SetIntraOpNumThreads(1);
+    if (gpu_mode != no_gpu) {
+      // detect if there is GPU
+      int devices = 0;
+      auto status = cudaGetDeviceCount(&devices);
+      bool is_gpu_available = (status == cudaSuccess && devices > 0);
+      if (is_gpu_available) {
+        // https://www.onnxruntime.ai/docs/reference/execution-providers/CUDA-ExecutionProvider.html
+        OrtCUDAProviderOptions options;
```
Comment (makortel) [inline PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:91]: I see in https://github.com/cms-sw/cmsdist/blob/f2fd150c7f36845f4f5a00a60b24dd33d265ec7b/onnxruntime.spec#L12-L14 and https://github.com/cms-sw/cmsdist/blob/f2fd150c7f36845f4f5a00a60b24dd33d265ec7b/onnxruntime.spec#L23-L27
```spec
%if "%{cmsos}" != "slc7_aarch64"
Requires: cuda cudnn
%endif
...
%if "%{cmsos}" != "slc7_aarch64"
USE_CUDA=ON
%else
USE_CUDA=OFF
%endif
```
that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)). What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc [hunk 4]
```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
 
   ONNXRuntime::~ONNXRuntime() {}
 
+  SessionOptions ONNXRuntime::defaultSessionOptions(GPUMode gpu_mode) {
+    SessionOptions sess_opts;
+    sess_opts.SetIntraOpNumThreads(1);
+    if (gpu_mode != no_gpu) {
+      // detect if there is GPU
+      int devices = 0;
+      auto status = cudaGetDeviceCount(&devices);
```
Comment (makortel) [inline PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:87]: I wonder if `CUDAService` should be used here instead for figuring our whether CUDA can be used or not?

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc [hunk 5]
```diff
@@ -78,6 +78,35 @@ namespace cms::Ort {
...
+        options.device_id = 0;
+        options.arena_extend_strategy = 0;
+        options.cuda_mem_limit = std::numeric_limits<std::size_t>::max();
+        options.cudnn_conv_algo_search = OrtCudnnConvAlgoSearch::EXHAUSTIVE;
+        options.do_copy_in_default_stream = 1;
+        sess_opts.AppendExecutionProvider_CUDA(options);
+      } else {
+        // if GPU is not available
+        if (gpu_mode == force_gpu) {
+          throw cms::Exception("RuntimeError") << "No GPU detected, cannot run ONNXRuntime on GPU.";
```
Comment (makortel) [inline PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:101]: This would be better as
```suggestion
          throw cms::Exception("UnavailableAccelerator") << "No GPU detected, cannot run ONNXRuntime on GPU.";
```
to align with #36699 (the string category should get mapped to `edm::errors::UnavailableAccelerator`) when #36699 gets in).

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 3 above)
Comment (makortel) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:91]: Given the condition `Requires: cuda` it would be better to make this package to explicitly depend on `cuda` external.

**Review by makortel — COMMENTED** [review]
Question beyond this PR: I'm wondering how ONNX supports (or plans to support) AMD and Intel GPUs? Are they going to bundle the support for all into a single dynamic library (at least what the "frontend" is concerned)? Or would there be one `onnxruntime` build for each GPU vendor? Or something else?

**Comment by fwyzard** [comment]
enable gpu

**Comment by fwyzard** [comment]
please test

**Comment by tvami** [comment]
@cmsbuild please test with https://github.com/cms-sw/cmssw/pull/36972

**Comment by hqucms** [comment]
> What NVIDIA GPU architectures (SM 6.0, 7.0, 7.5, _etc._) does ONNX support ?

I think it's listed [here](https://github.com/cms-externals/onnxruntime/blob/cms/v1.7.2/cmake/CMakeLists.txt#L1190-L1203) which seems to go back to even SM 3.7.

> Is it tied to how we use CUDA in CMSSW, or is it independent from it ?

It's based on the CUDA and cuDNN libraries in CMSSW.

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 4 above)
Comment (hqucms) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:87]: It should be used in the `EDModule` or directly here?

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 3 above)
Comment (hqucms) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:91]: > that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)). 

The restriction here should be `aarch64` and not really `slc7` I think:)

> What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?

The classes should be available at compile time but I guess the `AppendExecutionProvider_CUDA` call will fail at run time.

Comment (makortel) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:91]: > > that the CUDA backend is not supported for all of our platforms (even if slc7 is going to go away soonish(?)).
> 
> The restriction here should be `aarch64` and not really `slc7` I think:)

Thanks, then we should fix that :) Let me tag @smuzaffar here too.

> > What happens with these options if CUDA has been disabled in ONNX? Like, are the classes available at compile time?
> 
> The classes should be available at compile time but I guess the `AppendExecutionProvider_CUDA` call will fail at run time.

Ok, then this code could be fine as a first attempt (I'd expect it would need to be changed at some point, but hard to guess now how), assuming `PhysicsTools/ONNXRuntime` would gain explicit dependence on `cuda` external.

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 4 above)
Comment (makortel) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:87]: `CUDAService` can be used directly here. (this function is called from the EDModule' s, right? e.g. from their constructor)

File: PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc (hunk 3 above)
Comment (smuzaffar) [reply PhysicsTools/ONNXRuntime/src/ONNXRuntime.cc:91]: @hqucms @makortel , problem is with cudnn distribution for aarch64 for slc7. We use pre-build cudnn distributiuon ( https://github.com/cms-sw/cms-bot/blob/master/pr_testing/run-pr-comparisons#L124 ) . This was built for newer kernel ( and works on cs8) but failed for run/package for slc7. That is why the `cuda cudnn` is not included for `slc7_aarch64`.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
+reconstruction
- technical, no reco-related changes (new CUDA options for ONNXRuntime)

**Comment by jpata** [comment]
could this still enter pre6?

**Comment by perrotta** [comment]
> could this still enter pre6?

In principle, no:
- We are starting building pre6 kind of now
- This PR was not requested for pre6 during the last ORP, two days ago
- It is not even fully signed

Then, if there are really convinced that this is needed for 12_3_X, please justify it here and we can even decide to delay that build by half a day (and please try also to convince also @cms-sw/heterogeneous-l2 to sign)

**Comment by makortel** [comment]
@cmsbuild, please test for slc7_aarch64_gcc11

**Comment by makortel** [comment]
I'd prefer to have the use of `CUDAService` in `ONNXRuntime::defaultSessionOptions()` be done in this PR.

But maybe at this stage it would be safer to merge this after 12_3_0_pre6, and backport to 12_3_X after a few successful IBs (including at least one build with ARM and PPC)?

**Comment by perrotta** [comment]
> I'd prefer to have the use of `CUDAService` in `ONNXRuntime::defaultSessionOptions()` be done in this PR.
> 
> But maybe at this stage it would be safer to merge this after 12_3_0_pre6, and backport to 12_3_X after a few successful IBs (including at least one build with ARM and PPC)?

I think it is a fair proposal.
Backport of this PR can certainly be approved, once merged and tested in the new master.

**Comment by jpata** [comment]
no need to delay the build of course, I was just wondering

**Comment by hqucms** [comment]
> I'd prefer to have the use of `CUDAService` in `ONNXRuntime::defaultSessionOptions()` be done in this PR.
> 
> But maybe at this stage it would be safer to merge this after 12_3_0_pre6, and backport to 12_3_X after a few successful IBs (including at least one build with ARM and PPC)?

Sorry for the delay on this PR, I was occupied by a series urgent things this periods...

@makortel I thought a bit about the use of `CUDAService` -- I feel it would probably be better to use that in the EDModules, while keeping only a minimal dependency of ONNXRuntime package on other CMSSW packages -- so the ONNXRuntime package only depends on the external `onnxruntime` and `FWCore`. I think that allows a better separation of the packages and leaves more flexibility for various use cases of ONNXRuntime (so it could be used even outside EDModules as well if needed). 

We can even remove the CUDA availability detection in ONNXRuntime, so it purely gets the flag from outside (and using CUDAService in the EDModules).

What do you think?

**Comment by makortel** [comment]
@hqucms In general I'm in favor of keeping dependencies limited. I guess this could imply e.g. removing the `auto_gpu` option from the `GPUMode` enumeration, i..e the caller fully specifies the behavior.

It would make the calling code (EDModule) a bit more complex, but maybe that's ok. In principle one could abstract the checking logic to a helper function, but then we'd have a question of where to place that helper function if not here.

About

> (so it could be used even outside EDModules as well if needed)

I just want to note that the dependence on MessageLogger makes that a bit more difficult (for which an easiest, but not necessarily prettiest, solution would be to have `#ifdef` for the standalone case; probably best to be left at the time when such use is needed).

The ARM test shows that the use of CUDA Runtime API doesn't work (transitively) when ONNX wasn't compiled with CUDA support. From the discussion above those calls would be removed anyway. I wonder if `AppendExecutionProvider_CUDA()` is available at compile time in such case?

**Comment by vlimant** [comment]
The code overhead for the calling module looks minimal to me, and hence we should remove the detection from within ONNXRuntime ; if that's the show-stopper for this to move forward.

**Comment by vlimant** [comment]
The code overhead for the calling module looks minimal to me, and hence we should remove the detection from within ONNXRuntime ; if that's the show-stopper for this to move forward.

**Comment by hqucms** [comment]
Sorry for the long delay on this. I have implemented the proposed changes and removed all the CUDA detection part.

> The ARM test shows that the use of CUDA Runtime API doesn't work (transitively) when ONNX wasn't compiled with CUDA support. From the discussion above those calls would be removed anyway. I wonder if AppendExecutionProvider_CUDA() is available at compile time in such case?

I think it's available at compile time, but let's test it to be sure.

**Comment by makortel** [comment]
@cmsbuild, please test for slc7_aarch64_gcc11

File: PhysicsTools/ONNXRuntime/interface/ONNXRuntime.h [hunk 6]
```diff
@@ -29,6 +29,8 @@ namespace cms::Ort {
     ONNXRuntime& operator=(const ONNXRuntime&) = delete;
     ~ONNXRuntime();
 
+    static ::Ort::SessionOptions defaultSessionOptions(bool use_cuda = false);
```
Comment (makortel) [inline PhysicsTools/ONNXRuntime/interface/ONNXRuntime.h:32]: Maybe an enumeration (`enum class`) would be more self-documenting and extensible?

Comment (hqucms) [reply PhysicsTools/ONNXRuntime/interface/ONNXRuntime.h:32]: @makortel So something like:

```c++
enum class OrtSessOpt {
  cpu,
  cuda,
};
```

Comment (makortel) [reply PhysicsTools/ONNXRuntime/interface/ONNXRuntime.h:32]: Yeah, something like that.

**Comment by jpata** [comment]
@cmsbuild please test

**Comment by jpata** [comment]
Should we aim to have this in the prerelease next week? Is there still something pending here?
Also pinging the ML-prod team @riga @yongbinfeng on the timeline of this feature.

**Comment by fwyzard** [comment]
> There is my suggestion to use `enum class` instead of `bool` for denoting where to run [#36963 (comment)](https://github.com/cms-sw/cmssw/pull/36963#discussion_r833800258), that I think would be easier to address now rather than later.

-heterogeneous

**Comment by hqucms** [comment]
@makortel The `enum class` is used implemented in https://github.com/cms-sw/cmssw/pull/36963/commits/5eb31fbe030dc87b4e19fa4501d44b3b7d5c3560.
Sorry for the long delay, this somehow slipped through the cracks...

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by makortel** [comment]
+1

Thanks @hqucms

**Comment by jpata** [comment]
+reconstruction
- technical, readiness for CUDA use in ONNXRuntime
- no changes

**Comment by qliphy** [comment]
+1
